src/main/java/uk/gov/hmcts/reform/blobrouter/services/storage/BulkScanSasTokenCache.java

```diff
@@ -0,0 +1,93 @@
+package uk.gov.hmcts.reform.blobrouter.services.storage;
+
+import com.azure.storage.common.implementation.Constants;
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.Expiry;
+import org.checkerframework.checker.index.qual.NonNegative;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.blobrouter.clients.bulkscanprocessor.BulkScanProcessorClient;
+import uk.gov.hmcts.reform.blobrouter.exceptions.InvalidSasTokenException;
+
+import java.time.OffsetDateTime;
+import java.time.ZoneOffset;
+import java.time.temporal.TemporalAccessor;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+
+import static com.azure.storage.common.implementation.Constants.UrlConstants.SAS_EXPIRY_TIME;
+import static com.azure.storage.common.implementation.StorageImplUtils.parseQueryString;
+import static java.time.temporal.ChronoField.INSTANT_SECONDS;
+import static uk.gov.hmcts.reform.blobrouter.exceptions.InvalidSasTokenException.EXPIRY_NOT_FOUND;
+
+@Service
+public class BulkScanSasTokenCache {
+
+    private final BulkScanProcessorClient bulkScanSasTokenClient;
+    private final long refreshSasBeforeExpiry;
+
+    //key= container name, value = sastoken
+    private static Cache<String, String> tokenCache;
+
+    public BulkScanSasTokenCache(
+        BulkScanProcessorClient bulkScanSasTokenClient,
+        @Value("${bulk-scan-cache.refresh-before-expire-in-sec}") long refreshSasBeforeExpiry
+    ) {
+        this.bulkScanSasTokenClient = bulkScanSasTokenClient;
+        this.refreshSasBeforeExpiry = refreshSasBeforeExpiry;
+        tokenCache = Caffeine.newBuilder()
+            .expireAfter(new BulkScanSasTokenCacheExpiry())
+            .build();
+    }
+
+    public String getSasToken(String containerName) {
+        return tokenCache.get(containerName, conName -> this.createSasToken(conName));
```

Comment:
both are container name based. i would go with method reference on second param but that's just personal preference

Reply:
changed